nearform/labs-authorization
# Custom test framework

I opened this PR against my other PR so that changes to the code are clearer and it's hopefully easier to check if the new framework brings more clarity.

This PR currently contains the first step of such a framework: a function to build all initial records in one place. It is currently called `prepareTest` and returns an object intended to be later used to access created records (and in further iteration also dry-up the test definitions).

Two thing to note:
- Created records can be accessed by the key used in the records definition, but will have generic auto-generated ids
- Internally it still uses `lab.before` and `lab.after`, which I think is good for keeping everything in sync

At this time I'm seeking feedbacks on the type of approach and the "api definition" of the `prepareTest` function, basically this piece of code:

```js
      const test = prepareTest(lab, {
        teams: {
          calledTeam: { name: 'called team', organizationId, users: ['called'] }
        },
        users: {
          caller: { name: 'caller', organizationId, policies: ['testedPolicy'] },
          called: { name: 'called', organizationId }
        },
        policies: {
          testedPolicy: Policy()
        }
       })
```

Please go light on the implementation has it is still temporary to demonstrate my intentions (it will probably be moved to utils)

Next steps will be:
- use `prepareTest` in all experiments in this file, expand as needed
- create framework to define auth tests, something along the lines of
```js
      test('should authorize caller with policy for specific users')
        .withPolicy([{
          Effect: 'Allow',
          Action: ['authorization:users:read'],
          Resource: [`/authorization/user/WONKA/*/${calledId}`]
        }])
        .endpoint({
          method: 'GET',
          url: `/authorization/users/${calledId}`,
          headers: { authorization: callerId }
        })
        .shouldRespond(200)
```
